### PR TITLE
Basic testing for sandbox

### LINKED ISSUE #895: Use compatibility constraints from packages listed in extras
I was hoping that the `[compat]` constraint in `TestEnv` would constrain the versions I could install in `GPUEnv`.
The use-case where this would be interesting for is `Requires.jl`. `TestEnv` uses `Requires.jl` to load code that is dependent on a particular version of `CUDAnative`.

Alternatively instead of having a `extra` section one could have an `optional` section that has these stricter semantics.

```
name = "TestEnv"
uuid = "f59c4fb8-e42d-11e8-39b9-5d72c5061eb3"
version = "0.1.0"

[deps]
Requires = "ae029012-a4dd-5104-9daa-d747884805df"

[compat]
CUDAnative = "0.7"

[extras]
CUDAnative = "be33ccc6-a3ff-5ff2-a52e-74243cff1e17"
```

```
name = "GPUEnv"
uuid = "9eb97ce2-e42e-11e8-3709-85c2032a41b3"
version = "0.1.0"

[deps]
CUDAnative = "be33ccc6-a3ff-5ff2-a52e-74243cff1e17"
TestEnv = "f59c4fb8-e42d-11e8-39b9-5d72c5061eb3"
```

### LINKED ISSUE #1030: test resolves from wrong manifest
I added `Example` to `Literate/test/Project.toml` (and no `test/Manifest.toml` file):
```
(v1) pkg> st
    Status `~/.julia/environments/v1/Project.toml`
  [7876af07] Example v0.5.0
  [98b081ad] Literate v1.0.4 [`~/dev/Literate`]

(v1) pkg> test Literate
[...]
 Resolving package versions...
  Updating `/tmp/tmpE0z0sa/Project.toml`
  [7876af07] + Example v0.5.1            <-- Should be v0.5.0 from my v1 environment.
  [8dfed614] + Test
```

### PR DESCRIPTION
**Pull request by 00vareladavid**
- [x] fall back to parent dependency graph
- [x] test `test` sandboxing
- [x] test `build` sandboxing (very basic)

This will also test that the `compat` section of the sandbox project file is honored.

fix #1030, fix #895

### CODE REVIEW

File: src/Operations.jl [hunk 1]
```diff
@@ -1100,7 +1100,7 @@ end
 function with_temp_env(fn::Function, temp_env::String)
     cache = Base.active_project()
     try
-        Pkg.API.activate(temp_env)
+        Pkg.API.activate(temp_env; silent=true)
```
Comment (fredrikekre) [inline src/Operations.jl]: Wouldn't it be better to just `Base.ACTIVE_PROJECT[] = temp_env` instead of introducing a `silent` option?

Comment (StefanKarpinski) [reply src/Operations.jl]: That should have the same effect 👍

**Comment by 00vareladavid** [comment]
This will also test that the `compat` section of the sandbox project file is honored.

**Comment by 00vareladavid** [comment]
bors try

**Comment by 00vareladavid** [comment]
bors try

File: src/API.jl [hunk 2]
```diff
@@ -139,15 +140,15 @@ up(pkgs::Vector{PackageSpec}; kwargs...)               = up(Context(), pkgs; kwa
 
 function up(ctx::Context, pkgs::Vector{PackageSpec};
             level::UpgradeLevel=UPLEVEL_MAJOR, mode::PackageMode=PKGMODE_PROJECT,
-            do_update_registry=true, kwargs...)
+            update_registry::Bool=true, skip_check::Bool=true, kwargs...)
```
Comment (fredrikekre) [inline src/API.jl]: `skip_check` seems to be unused.

File: src/API.jl [hunk 3]
```diff
@@ -81,6 +81,7 @@ function add(ctx::Context, pkgs::Vector{PackageSpec}; kwargs...)
     end
 
     ctx.preview && preview_info()
+    Types.update_registries(ctx)
```
Comment (fredrikekre) [inline src/API.jl]: Did this happen further down in the call chain before?

File: src/API.jl [hunk 4]
```diff
@@ -34,7 +34,8 @@ develop(pkg::Union{AbstractString, PackageSpec}; kwargs...) = develop([pkg]; kwa
 develop(pkgs::Vector{<:AbstractString}; kwargs...) =
     develop([check_package_name(pkg, :develop) for pkg in pkgs]; kwargs...)
 develop(pkgs::Vector{PackageSpec}; kwargs...)      = develop(Context(), pkgs; kwargs...)
-function develop(ctx::Context, pkgs::Vector{PackageSpec}; shared::Bool=true, kwargs...)
+function develop(ctx::Context, pkgs::Vector{PackageSpec};
+                 shared::Bool=true, keep_manifest::Bool=false, kwargs...)
```
Comment (fredrikekre) [inline src/API.jl]: What does `keep_manifest` do? Which manifest is kept?

File: src/API.jl [hunk 5]
```diff
@@ -224,32 +225,28 @@ function free(ctx::Context, pkgs::Vector{PackageSpec}; kwargs...)
...
-test(pkgs::Vector{PackageSpec}; kwargs...)        = test(Context(), pkgs; kwargs...)
-
-function test(ctx::Context, pkgs::Vector{PackageSpec}; coverage=false, kwargs...)
-    pkgs = deepcopy(pkgs)  # deepcopy for avoid mutating PackageSpec members
+test(;kwargs...)                                         = test(PackageSpec[]; kwargs...)
+test(pkg::Union{AbstractString, PackageSpec}; kwargs...) = test([pkg]; kwargs...)
+test(pkgs::Vector{<:AbstractString}; kwargs...)          = test([PackageSpec(pkg) for pkg in pkgs]; kwargs...)
+test(pkgs::Vector{PackageSpec}; kwargs...)               = test(Context(), pkgs; kwargs...)
+function test(ctx::Context, pkgs::Vector{PackageSpec};
+              coverage=false, test_fn=nothing, kwargs...)
```
Comment (fredrikekre) [inline src/API.jl]: Why is `test_fn` needed?

File: src/Operations.jl [hunk 6]
```diff
@@ -1112,32 +1121,64 @@ function gen_test_code(testfile::String; coverage=false)
...
-    cache = Base.active_project()
+    old_load_path = copy(LOAD_PATH)
+    old_active_project = Base.ACTIVE_PROJECT[]
     try
-        Pkg.API.activate(temp_env)
-        fn()
+        withenv("JULIA_LOAD_PATH" => temp_env) do
+            empty!(LOAD_PATH)
+            push!(LOAD_PATH, "@", "@v#.#", "@stdlib")
+            Base.ACTIVE_PROJECT[] = temp_env
```
Comment (fredrikekre) [inline src/Operations.jl]: Wouldn't it be better if this ran in the spawned process?

File: src/API.jl (hunk 3 above)
Comment (00vareladavid) [reply src/API.jl]: Yes, it was in `handle_repos_add`. It was the first call and we always call `handle_repos_add`. I thought it better if it was more clear where we were updating the registry

File: src/API.jl (hunk 5 above)
Comment (00vareladavid) [reply src/API.jl]: I used it for testing purposes. I allows one to hook into the `test` environment. We can also put it in `Context` or have some other way of testing within the `test` environment. I just needed something to be in place: I'm not super convinced of a superior alternative.

File: src/API.jl (hunk 2 above)
Comment (00vareladavid) [reply src/API.jl]: Yup, I forgot to remove it :+1:

File: src/Operations.jl (hunk 6 above)
Comment (00vareladavid) [reply src/Operations.jl]: In the process we spawn for `test`? In what way would it be better? Are you referring to the whole `with_temp_env` function?

File: src/API.jl (hunk 4 above)
Comment (00vareladavid) [reply src/API.jl]: The manifest created here: https://github.com/JuliaLang/Pkg.jl/pull/1033/files#diff-920ed338dcbd9d62489c6d12dc2fbb18R1172. Together, its pretty much these two lines which support #1030

**Comment by 00vareladavid** [comment]
@fredrikekre This [test](https://github.com/JuliaLang/Pkg.jl/pull/1033/files#diff-60f2cbfa19264866ccaef20998742653R20)  represents the behavior we were discussing at #1030.

**Comment by 00vareladavid** [comment]
@fredrikekre most helpful here would be tests which break the sandbox or check edge cases. The code here is not super clean, but I think its better if we pin down behavior first.

File: test/test_packages/SandboxFallback2/Project.toml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+name = "Foo"
+uuid = "02250abe-2050-11e9-017e-b301a2b5bcc4"
+authors = ["David Varela <00.varela.david@gmail.com>"]
```
Comment (fredrikekre) [inline test/test_packages/SandboxFallback2/Project.toml]: Maybe remove this and the other `authors` fields?

Comment (00vareladavid) [reply test/test_packages/SandboxFallback2/Project.toml]: Sure, I always forget to do that.

Comment (StefanKarpinski) [reply test/test_packages/SandboxFallback2/Project.toml]: But he wants to leave his mark 😁

Comment (00vareladavid) [reply test/test_packages/SandboxFallback2/Project.toml]: Deezy wuz here

**Comment by 00vareladavid** [comment]
bors try

**Comment by fredrikekre** [comment]
:tada:
